Docker setup for React.js

landswellsong opened the pull request:
- First run is painfully slow, keep that in mind
- Dev mode syncs the `/app/` directory, do we need more of those?

Seems to work overall.

landswellsong committed:
Allowing to edit backend on the fly

landswellsong committed:
Temp code move port, ignore this

landswellsong committed:
Primitive nodejs dockerfile

landswellsong committed:
Pre-merger preparation

landswellsong committed:
Merge remote-tracking branch 'origin/reactjs' into docker_react

landswellsong committed:
Dockerfile for reactjs branch

landswellsong committed:
Merge remote-tracking branch 'origin/reactjs' into docker_react

landswellsong committed:
Operable version with dynamic deployment

landswellsong committed:
Production mode

landswellsong committed:
Syncing the /app/ directory in dev mode

landswellsong commented on conf/containers.yml:
```diff
@@ -36,10 +39,14 @@
   depend: [ "${conf.hostname_api}" ]
+  reload:
+    # TODO: do we need other directories synced?
+    - [ "${conf.dir_web}/app", "/usr/src/promis/app" ]
```
Add stuff here to sync other folders as well. Avoid syncing the whole frontend directory since it will overmount `node_modules` on the container.

landswellsong commented on frontend/Dockerfile:
```diff
@@ -17,7 +19,18 @@ ADD nginx.conf /etc/nginx/
+# Add the actual app
+ADD run_web_promis.sh /
+ADD app /usr/src/promis/app/
+# TODO: please remove what's no longer applicable!
```
Add(remove) stuff here if you need more (less) source dirs copied over to the container as the part of a normal build process.

landswellsong commented on frontend/promis.conf:
```diff
@@ -31,6 +32,7 @@ server {
 
     # Django's static files (for backend's admin)
     location /static/ {
+        ${conf.web_root_config}
```
Add more things like this if you need other paths conditionally handled too (i.e. switched between node.js server in dev and static files in production)

landswellsong committed:
Travis now runs data_updates and web.promis startup in parallel

landswellsong committed:
fixes #120

0181532686cf4a31163be0bf3e6bb6732bf commented on frontend/webpack.config.js:
```diff
@@ -12,7 +12,7 @@ var path = require('path');
     entry: path.join(__dirname, 'app', 'index.js'),
     output: {
-        path: path.resolve(__dirname, 'public'),
+        path: '/var/www/promis', // TODO: do we need to parametrise this?
```
`path.join` should just put it to the right place

0181532686cf4a31163be0bf3e6bb6732bf replied:
I mean if this script is located in `/var/www/promis`, then `__dirname === '/var/www/promis'` and `path.join(__dirname, 'public')` should be just fine

landswellsong replied:
No, this script is run at `/usr/src/promis`, not at `/var/www/promis` which should only contain ouptut. Also why do we need `public` directory in first place? 

My reasoning here:
- `/usr/src/promis` ⇒ source React files
- `/var/www/promis` ⇒ compiled JS/HTML/whatev that is served by nginx, no development files there

Is that correct?

0181532686cf4a31163be0bf3e6bb6732bf committed:
Merge branch 'reactjs' into docker_react

0181532686cf4a31163be0bf3e6bb6732bf committed:
Update promis.conf

Add user profile endpoint

landswellsong committed:
Nginx now serves from /usr/src/promis/public folder

0181532686cf4a31163be0bf3e6bb6732bf committed:
Missed comma

0181532686cf4a31163be0bf3e6bb6732bf committed:
Removed assets

landswellsong committed:
Move to a faster container

landswellsong committed:
Removed the weird magic command

Gives 0.05% speedup and 500% increase in variance based on 3+3 tests, don't think it's affecting things that much
